File: 949.Largest-Time-for-Given-Digits.py
# algorithms
# Easy (30.52%)
# Total Accepted:    3,167
# Total Submissions: 10,376
# beats 100.0% of python submissions


class Solution(object):
    def largestTimeFromDigits(self, A):
        """
        :type A: List[int]
        :rtype: str
        """
        hash_map = {}
        for num in A:
            if num in hash_map:
                hash_map[num] = hash_map[num] + 1
            else:
                hash_map[num] = 1

        hour_high = [i for i in xrange(2, -1, -1)]
        hour_low = [i for i in xrange(3, -1, -1)]
        reverse_order = [i for i in xrange(9, -1, -1)]
        minute_high = [i for i in xrange(5, -1, -1)]

        res = [[], False]

        def resursive(cnt_idx, path):
            if res[1]:
                return
            if cnt_idx == 4:
                res[0] = path
                res[1] = True
                return

            arr = None
            if cnt_idx == 0:
                arr = hour_high
            elif cnt_idx == 1:
                arr = hour_low if path[0] == '2' else reverse_order
            elif cnt_idx == 2:
                arr = minute_high
            else:
                arr = reverse_order

            for i in arr:
                if hash_map.get(i, 0) > 0:
                    hash_map[i] = hash_map[i] - 1
                    resursive(cnt_idx + 1, path + [str(i)])
                    hash_map[i] = hash_map[i] + 1

        resursive(0, [])

        if len(res[0]) == 0:
            return ''

        res[0].insert(2, ':')

        return ''.join(res[0])

File: 949.Largest-Time-for-Given-Digits.java
// https://leetcode.com/problems/largest-time-for-given-digits/
//
// algorithms
// Easy (34.8%)
// Total Accepted: 11.8K
// Total Submissions: 33.9K


class Solution {

    private static String res;
    private static final int[] UPPER = new int[] { 2, 3, 5, 9 };

    public String largestTimeFromDigits(int[] A) {
        res = null;

        Arrays.sort(A);
        int[] path = new int[4];
        boolean[] isVisited = new boolean[4];

        recursive(A, path, isVisited, 0);

        return res == null ? "" : res;
    }

    private void recursive(int[] A, int[] path, boolean[] isVisited, int timeIdx) {
        if (res != null) {
            return;
        }

        if (timeIdx == 4) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < 2; i++) {
                sb.append(path[i]);
            }
            sb.append(':');
            for (int i = 2; i < 4; i++) {
                sb.append(path[i]);
            }
            res = sb.toString();
            return;
        }

        for (int i = 3; i >= 0; i--) {
            if (!isVisited[i]) {
                if (timeIdx == 1) {
                    if (path[0] == 2 && A[i] > UPPER[timeIdx]) {
                        continue;
                    }
                } else if (A[i] > UPPER[timeIdx]) {
                    continue;
                }
                isVisited[i] = true;
                path[timeIdx] = A[i];
                recursive(A, path, isVisited, timeIdx + 1);
                isVisited[i] = false;
            }
        }
    }

}